lib/buildkite/test_collector/session.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module Buildkite::TestCollector
+  class Session
+    UPLOAD_THREAD_TIMEOUT = 60
+    UPLOAD_SESSION_TIMEOUT = 60
+    UPLOAD_API_MAX_RESULTS = 5000
+
+    def initialize
+      @send_queue_ids = []
+      @upload_threads = []
+    end
+
+    def add_example_to_send_queue(id)
+      @send_queue_ids << id
+
+      if @send_queue_ids.size >= Buildkite::TestCollector.batch_size
+        send_ids = @send_queue_ids.shift(Buildkite::TestCollector.batch_size)
+        upload_data(send_ids)
+      end
+    end
+
+    def send_remaining_data
+      return if @send_queue_ids.empty?
+
+      upload_data(@send_queue_ids)
+    end
+
+    def close
+      # There are two thread joins here, because the inner join will wait up to
+      # UPLOAD_THREAD_TIMEOUT seconds PER thread that is uploading data, i.e.
+      # n_threads x UPLOAD_THREAD_TIMEOUT latency if Buildkite happens to be
+      # down. By wrapping that in an outer thread join with the
+      # UPLOAD_SESSION_TIMEOUT, we ensure that we only wait a max of
+      # UPLOAD_SESSION_TIMEOUT seconds before the session exits.
+      Thread.new do
+        @upload_threads.each { |t| t.join(UPLOAD_THREAD_TIMEOUT) }
+      end.join(UPLOAD_SESSION_TIMEOUT)
+
+      @upload_threads.each { |t| t&.kill }
```

Comment:
What's the reason behind the lonely operator? Do we expect nils in `@upload_threads`?

Nitpick - can we kill the threads if `join` times out and returns nil on line 37?

Reply:
I had the lonely operator in there cos I thought you couldn't kill a thread that was already dead (i.e. the upload has already been sent) but it seems you can call kill on a dead thread still so I guess there's no reason to keep the lonely operator there. It's also safe to kill a thread that has timed out its join